File: collect_app/src/test/java/org/odk/collect/android/audio/AudioControllerViewTest.java
package org.odk.collect.android.audio;

import android.widget.SeekBar;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.odk.collect.android.R;
import org.odk.collect.android.support.SwipableParentActivity;
import org.robolectric.RobolectricTestRunner;

import static org.hamcrest.Matchers.equalTo;
import static org.junit.Assert.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.odk.collect.android.support.RobolectricHelpers.buildThemedActivity;
import static org.robolectric.Shadows.shadowOf;
import static org.robolectric.shadows.ShadowView.innerText;

@RunWith(RobolectricTestRunner.class)
public class AudioControllerViewTest {

    private SwipableParentActivity activity;
    private AudioControllerView view;

    @Before
    public void setup() {
        activity = buildThemedActivity(SwipableParentActivity.class).get();
        view = new AudioControllerView(activity);
    }

    @Test
    public void clickingFastForward_whenPostionAtDuration_skipsToDuration() {
        AudioControllerView.Listener listener = mock(AudioControllerView.Listener.class);

        view.setListener(listener);
        view.setDuration(1000);
        view.setPosition(1000);

        view.findViewById(R.id.fastForwardBtn).performClick();

        assertThat(innerText(view.findViewById(R.id.currentDuration)), equalTo("00:01"));
        verify(listener).onPositionChanged(1000);
    }

    @Test
    public void clickingFastRewind_whenPostionAtZero_skipsTo0() {
        AudioControllerView.Listener listener = mock(AudioControllerView.Listener.class);

        view.setListener(listener);
        view.setDuration(1000);
        view.setPosition(0);

        view.findViewById(R.id.fastRewindBtn).performClick();

        assertThat(innerText(view.findViewById(R.id.currentDuration)), equalTo("00:00"));
        verify(listener).onPositionChanged(0);
    }

    @Test
    public void whenSwiping_notifiesSwipableParent() {
        SeekBar seekBar = view.findViewById(R.id.seekBar);

        shadowOf(seekBar).getOnSeekBarChangeListener().onStartTrackingTouch(seekBar);
        assertThat(activity.isSwipingAllowed(), equalTo(false));

        shadowOf(seekBar).getOnSeekBarChangeListener().onStopTrackingTouch(seekBar);
        assertThat(activity.isSwipingAllowed(), equalTo(true));
    }
}

File: collect_app/src/test/java/org/odk/collect/android/audio/AudioControllerViewTest.kt
package org.odk.collect.android.audio

import android.widget.SeekBar
import androidx.test.ext.junit.runners.AndroidJUnit4
import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito
import org.mockito.Mockito.verify
import org.mockito.Mockito.verifyNoInteractions
import org.mockito.Mockito.verifyNoMoreInteractions
import org.odk.collect.android.R
import org.odk.collect.android.support.CollectHelpers
import org.odk.collect.android.support.SwipableParentActivity
import org.robolectric.Shadows
import org.robolectric.shadows.ShadowView

@RunWith(AndroidJUnit4::class)
class AudioControllerViewTest {
    private lateinit var activity: SwipableParentActivity
    private lateinit var view: AudioControllerView

    @Before
    fun setup() {
        activity = CollectHelpers.buildThemedActivity(
            SwipableParentActivity::class.java
        ).get()
        view = AudioControllerView(activity)
    }

    @Test
    fun `setDuration() shows duration in minutes and seconds`() {
        view.setDuration(52000)

        assertThat(
            view.binding.totalDuration.text.toString(),
            equalTo("00:52")
        )
    }

    @Test
    fun `setPosition() shows position in minutes and seconds`() {
        view.setDuration(65000)
        view.setPosition(64000)

        assertThat(
            view.binding.currentDuration.text.toString(),
            equalTo("01:04")
        )
    }

    @Test
    fun `setPosition() changes seekBar position`() {
        view.setDuration(65000)

        assertThat(view.binding.seekBar.progress, `is`(0))
        assertThat(view.binding.seekBar.max, `is`(65000))

        view.setPosition(8000)

        assertThat(view.binding.seekBar.progress, Matchers.`is`(8000))
        assertThat(view.binding.seekBar.max, Matchers.`is`(65000))
    }

    @Test
    fun `swiping seekBar when paused skips to position once stopped`() {
        val listener = Mockito.mock(
            AudioControllerView.Listener::class.java
        )

        view.setListener(listener)
        view.setDuration(12000)

        val seekBar = view.binding.seekBar

        Shadows.shadowOf(seekBar).onSeekBarChangeListener.onStartTrackingTouch(seekBar)
        Shadows.shadowOf(seekBar).onSeekBarChangeListener.onProgressChanged(seekBar, 7000, true)

        assertThat(
            ShadowView.innerText(view.findViewById(R.id.currentDuration)),
            equalTo("00:07")
        )

        verifyNoInteractions(listener) // We don't change position yet

        Shadows.shadowOf(seekBar).onSeekBarChangeListener.onProgressChanged(seekBar, 5000, true)

        assertThat(
            ShadowView.innerText(view.findViewById(R.id.currentDuration)),
            equalTo("00:05")
        )
        verifyNoInteractions(listener) // We don't change position yet

        Shadows.shadowOf(seekBar).onSeekBarChangeListener.onStopTrackingTouch(seekBar)

        assertThat(
            ShadowView.innerText(view.findViewById(R.id.currentDuration)),
            equalTo("00:05")
        )

        verify(listener).onPositionChanged(5000)
    }

    @Test
    fun `swiping seekBar when playing pauses and then skips to position and plays once stopped`() {
        val listener = Mockito.mock(
            AudioControllerView.Listener::class.java
        )
        view.setListener(listener)
        view.setDuration(12000)
        view.setPlaying(true)

        val seekBar = view.binding.seekBar

        Shadows.shadowOf(seekBar).onSeekBarChangeListener.onStartTrackingTouch(seekBar)

        verify(listener).onPauseClicked()

        Shadows.shadowOf(seekBar).onSeekBarChangeListener.onProgressChanged(seekBar, 7000, true)

        assertThat(
            ShadowView.innerText(view.findViewById(R.id.currentDuration)),
            equalTo("00:07")
        )

        verifyNoMoreInteractions(listener) // We don't change position yet

        Shadows.shadowOf(seekBar).onSeekBarChangeListener.onProgressChanged(seekBar, 5000, true)

        assertThat(
            ShadowView.innerText(view.findViewById(R.id.currentDuration)),
            equalTo("00:05")
        )

        verifyNoMoreInteractions(listener) // We don't change position yet

        Shadows.shadowOf(seekBar).onSeekBarChangeListener.onStopTrackingTouch(seekBar)

        assertThat(
            ShadowView.innerText(view.findViewById(R.id.currentDuration)),
            equalTo("00:05")
        )

        verify(listener).onPositionChanged(5000)
        verify(listener).onPlayClicked()
    }

    @Test
    fun `when swiping notifies swipeable parent`() {
        val seekBar = view.findViewById<SeekBar>(R.id.seekBar)
        Shadows.shadowOf(seekBar).onSeekBarChangeListener.onStartTrackingTouch(seekBar)

        assertThat(activity.isSwipingAllowed, equalTo(false))

        Shadows.shadowOf(seekBar).onSeekBarChangeListener.onStopTrackingTouch(seekBar)

        assertThat(activity.isSwipingAllowed, equalTo(true))
    }
}
